### ISSUE TITLE
[Phase 3] Gaussian Mixture Segmentation

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/dynamic-clv-segmentation#4]
## Summary

Implement probabilistic customer segmentation using Gaussian Mixture Models with automatic K selection via BIC. Segments will inform segment-specific CLV regressors.

**Priority:** 🔴 Critical - Core component for stacked prediction
**Parent Issue:** #1
**Depends On:** Phase 1 (#2)

---

## Objectives

1. Build GMM wrapper with BIC-based model selection
2. Return soft (probabilistic) cluster assignments
3. Create segment profiler for business interpretability
4. Handle edge cases (small clusters, K=1)

---

## Files to Create

| File | Lines (est.) | Description |
|------|--------------|-------------|
| `src/segmentation/__init__.py` | 1 | Package init |
| `src/segmentation/gmm.py` | 180 | GMM with BIC selection |
| `src/segmentation/segment_profiler.py` | 100 | Segment interpretation |
| `tests/test_segmentation.py` | 150 | Unit tests |

---

### GMM Segmenter
- [ ] Implement K selection via BIC minimization
- [ ] Fit GMMs for K in range [k_min, k_max]
- [ ] Return soft probabilities (not just hard labels)
- [ ] Handle convergence issues gracefully
- [ ] Add K-Means++ initialization for stability

### Segment Profiler
- [ ] Compute per-segment feature means/stds
- [ ] Generate human-readable segment descriptions
- [ ] Identify distinguishing features per segment
- [ ] Support custom feature names

### Testing
- [ ] Verify BIC selects correct K on synthetic data
- [ ] Test soft assignments sum to 1.0
- [ ] Test edge case: all customers identical
- [ ] Test edge case: K=1 (no segments)

---

## Starter Code

📄 **File:** `src/segmentation/gmm.py`

```python
"""Gaussian Mixture Model segmentation with automatic K selection."""

from dataclasses import dataclass
from sklearn.mixture import GaussianMixture
import numpy as np


@dataclass
class SegmentationResult:
    """Container for segmentation outputs."""
    n_segments: int
    probabilities: np.ndarray  # (n_customers, n_segments)
    labels: np.ndarray  # Hard assignments
    bic_scores: dict[int, float]  # K -> BIC
    segment_profiles: dict[int, dict]  # Per-segment stats


class CustomerSegmenter:
    """
    Probabilistic segmentation using Gaussian Mixture Models.

    Automatically selects optimal K using BIC criterion
    (lower BIC = better model).
    """

    def __init__(
        self,
        k_range: tuple[int, int] = (2, 10),
        random_state: int = 42,
        n_init: int = 5,
        max_iter: int = 100
    ):
        self.k_range = k_range
        self.random_state = random_state
        self.n_init = n_init
        self.max_iter = max_iter

        self._best_model: GaussianMixture | None = None
        self._best_k: int | None = None
        self._bic_scores: dict[int, float] = {}

    def fit(self, X: np.ndarray) -> "CustomerSegmenter":
        """Fit GMMs for each K and select best by BIC."""
        best_bic = np.inf

        for k in range(self.k_range[0], self.k_range[1] + 1):
            gmm = GaussianMixture(
                n_components=k,
                random_state=self.random_state,
                n_init=self.n_init,
                max_iter=self.max_iter,
                init_params='k-means++'
            )
            gmm.fit(X)
            bic = gmm.bic(X)
            self._bic_scores[k] = bic

            if bic < best_bic:
                best_bic = bic
                self._best_model = gmm
                self._best_k = k

        return self

    def predict_proba(self, X: np.ndarray) -> np.ndarray:
        """Return soft cluster probabilities."""
        if self._best_model is None:
            raise ValueError("Must call fit() first")
        return self._best_model.predict_proba(X)

    def predict(self, X: np.ndarray) -> np.ndarray:
        """Return hard cluster labels."""
        if self._best_model is None:
            raise ValueError("Must call fit() first")
        return self._best_model.predict(X)

    def get_result(
        self,
        X: np.ndarray,
        feature_names: list[str] | None = None
    ) -> SegmentationResult:
        """Return full segmentation result with profiles."""
        probs = self.predict_proba(X)
        labels = self.predict(X)

        # Compute segment profiles
        profiles = {}
        for k in range(self._best_k):
            mask = labels == k
            profiles[k] = {
                "size": mask.sum(),
                "proportion": mask.mean(),
                "mean": X[mask].mean(axis=0).tolist(),
                "std": X[mask].std(axis=0).tolist(),
            }
            if feature_names:
                profiles[k]["feature_names"] = feature_names

        return SegmentationResult(
            n_segments=self._best_k,
            probabilities=probs,
            labels=labels,
            bic_scores=self._bic_scores,
            segment_profiles=profiles
        )
```

📄 **File:** `tests/test_segmentation.py`

```python
"""Unit tests for GMM segmentation."""

import pytest
import numpy as np
from src.segmentation.gmm import CustomerSegmenter


@pytest.fixture
def three_cluster_data():
    """Synthetic data with 3 well-separated clusters."""
    np.random.seed(42)
    n_per_cluster = 100

    c1 = np.random.randn(n_per_cluster, 4) + np.array([0, 0, 0, 0])
    c2 = np.random.randn(n_per_cluster, 4) + np.array([5, 5, 5, 5])
    c3 = np.random.randn(n_per_cluster, 4) + np.array([-5, 5, -5, 5])

    return np.vstack([c1, c2, c3])


def test_bic_selects_correct_k(three_cluster_data):
    """BIC should select K=3 for 3-cluster data."""
    segmenter = CustomerSegmenter(k_range=(2, 6))
    segmenter.fit(three_cluster_data)

    assert segmenter._best_k == 3


def test_soft_assignments_sum_to_one(three_cluster_data):
    """Probabilities should sum to 1.0 per sample."""
    segmenter = CustomerSegmenter(k_range=(2, 4))
    segmenter.fit(three_cluster_data)

    probs = segmenter.predict_proba(three_cluster_data)
    row_sums = probs.sum(axis=1)

    np.testing.assert_allclose(row_sums, 1.0, rtol=1e-5)


def test_segment_profiles_computed(three_cluster_data):
    """Result should include segment profiles."""
    segmenter = CustomerSegmenter(k_range=(2, 4))
    segmenter.fit(three_cluster_data)

    result = segmenter.get_result(three_cluster_data)

    assert len(result.segment_profiles) == result.n_segments
    for k, profile in result.segment_profiles.items():
        assert "size" in profile
        assert "mean" in profile
        assert profile["size"] > 0
```

---

## Technical Challenges

| Challenge | Mitigation |
|-----------|------------|
| BIC unstable with small N | Set minimum samples per component |
| Convergence failures | Multiple initializations, K-Means++ init |
| Degenerate clusters | Monitor cluster sizes, set min threshold |
| High-dimensional input | Consider PCA preprocessing |

---

## Definition of Done

- [ ] BIC correctly identifies K on synthetic data
- [ ] Soft probabilities sum to 1.0 per row
- [ ] Segment profiles provide actionable stats
- [ ] Handles K=1 gracefully
- [ ] All tests pass
